consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -57,26 +57,210 @@ enum Action {
+    fn proposal(&self) -> Option<&Proposal<D>> {
+        self.proposal.as_ref()
+    }
+
+    /// Returns the current [`ProposalStatus`] for the slot.
+    fn status(&self) -> ProposalStatus {
+        self.status
+    }
+
+    /// Returns `true` when the caller should begin constructing a proposal.
+    ///
+    /// This method does not mutate the slot; callers must invoke [`Self::set_building`]
+    /// once they commit to building so future calls to `should_build` short-circuit.
+    /// It returns `false` if a proposal already exists or [`Self::set_building`] has
+    /// been called.
+    fn should_build(&self) -> bool {
+        if self.requested_build || self.proposal.is_some() {
+            return false;
+        }
+        true
+    }
+
+    /// Records that proposal construction is in progress so subsequent calls to
+    /// [`Self::should_build`] return `false`.
+    fn set_building(&mut self) {
+        self.requested_build = true;
+    }
+
+    /// Returns `true` if verifying the proposal has already been requested.
+    fn has_requested_verify(&self) -> bool {
+        self.requested_verify
+    }
+
+    /// Attempts to request verification of the proposal.
+    ///
+    /// Returns `true` if the verify request is marked during this call, or
+    /// `false` if verification was already requested.
+    fn request_verify(&mut self) -> bool {
+        if self.requested_verify {
+            return false;
+        }
+        self.requested_verify = true;
+        true
+    }
+
+    /// Records the proposal that we produced locally.
+    ///
+    /// The slot must not already contain a proposal. Recording our proposal
+    /// immediately marks it as verified and ensures both build and verify
+    /// requests are set so future calls short-circuit.
+    fn record_our_proposal(&mut self, replay: bool, proposal: Proposal<D>) {
+        assert!(self.proposal.is_none() || replay, "proposal already set");
+        self.proposal = Some(proposal);
+        self.status = ProposalStatus::Verified;
+        self.requested_build = true;
+        self.requested_verify = true;
+    }
+
+    /// Promotes the slot back to [`ProposalStatus::Verified`] if it was unverified.
+    ///
+    /// Returns `true` if the status changed, or `false` if it was already in a
+    /// different state.
+    fn mark_verified(&mut self) -> bool {
+        if self.status != ProposalStatus::Unverified {
+            return false;
+        }
+        self.status = ProposalStatus::Verified;
+        true
+    }
+
+    /// Updates the slot with a proposal observed from the network or replay.
+    ///
+    /// Returns a [`ProposalChange`] describing how the slot was mutated.
+    ///
+    /// When `certificate` is `true`, the provided proposal is considered authoritative
+    /// and should override any conflicting votes that may have already been tracked.
+    fn update(&mut self, proposal: &Proposal<D>, certificate: bool) -> ProposalChange<D> {
```

Comment:
TODO: this won't work if we end up storing votes for multiple proposals in the journal (we'll hit Replaced on the first conflict and then refuse to update)